libraries/tools/kotlin-gradle-plugin/src/main/kotlin/org/jetbrains/kotlin/gradle/plugin/mpp/syncKotlinAndAndroidSourceSets.kt

```diff
@@ -0,0 +1,123 @@
+
+import com.android.build.gradle.BaseExtension
+import com.android.build.gradle.api.AndroidSourceSet
+import org.gradle.api.Project
+import org.jetbrains.kotlin.gradle.dsl.kotlinExtension
+import org.jetbrains.kotlin.gradle.internal.Kapt3GradleSubplugin
+import org.jetbrains.kotlin.gradle.plugin.*
+import org.jetbrains.kotlin.gradle.plugin.AbstractAndroidProjectHandler.Companion.kotlinSourceSetNameForAndroidSourceSet
+import org.jetbrains.kotlin.gradle.plugin.addConvention
+import java.io.File
+
+internal fun syncKotlinAndAndroidSourceSets(target: KotlinAndroidTarget) {
+    val project = target.project
+    val android = project.extensions.getByName("android") as BaseExtension
+
+    android.sourceSets.all { androidSourceSet ->
+        val kotlinSourceSetName = kotlinSourceSetNameForAndroidSourceSet(target, androidSourceSet.name)
+        val kotlinSourceSet = project.kotlinExtension.sourceSets.maybeCreate(kotlinSourceSetName)
+        createDefaultDependsOnEdges(target, kotlinSourceSet, androidSourceSet)
+        syncKotlinAndAndroidSourceDirs(target, kotlinSourceSet, androidSourceSet)
+        syncKotlinAndAndroidResources(target, kotlinSourceSet, androidSourceSet)
+        androidSourceSet.addConvention(KOTLIN_DSL_NAME, kotlinSourceSet)
+
+        ifKaptEnabled(project) {
+            Kapt3GradleSubplugin.createAptConfigurationIfNeeded(project, androidSourceSet.name)
+        }
+    }
+}
+
+private fun createDefaultDependsOnEdges(
+    target: KotlinAndroidTarget,
+    kotlinSourceSet: KotlinSourceSet,
+    androidSourceSet: AndroidSourceSet
+) {
+    val commonSourceSetName = when (androidSourceSet.name) {
+        "main" -> "commonMain"
+        "test" -> "commonTest"
+        "androidTest" -> "commonTest"
+        else -> return
+    }
+    val commonSourceSet = target.project.kotlinExtension.sourceSets.findByName(commonSourceSetName) ?: return
+    kotlinSourceSet.dependsOn(commonSourceSet)
+}
+
+private fun syncKotlinAndAndroidSourceDirs(
+    target: KotlinAndroidTarget, kotlinSourceSet: KotlinSourceSet, androidSourceSet: AndroidSourceSet
+) {
+    val disambiguationClassifier = target.disambiguationClassifier
+
+    /*
+    Mitigate ambiguity!
+    Example: disambiguationClassifier="android"
+    Source Directory "src/androidTest/kotlin"
+        -- could be claimed by kotlin {android}Test (unit test)
+        -- could be claimed by Android androidTest (instrumented test)
+
+    The Kotlin source set would win in this scenario.
+     */
+    if (disambiguationClassifier == null || !androidSourceSet.name.startsWith(disambiguationClassifier)) {
+        kotlinSourceSet.kotlin.srcDir("src/${androidSourceSet.name}/kotlin")
+    }
+
+    kotlinSourceSet.kotlin.srcDirs(*androidSourceSet.java.srcDirs.toTypedArray())
+    androidSourceSet.java.srcDirs(*kotlinSourceSet.kotlin.srcDirs.toTypedArray())
+
+    /*
+    Make sure to include user configuration as well.
+    Unfortunately, there does not exist any "ad-hoc" API like `all`.
```

Comment:
AGP Variant API should give you a hook that you can use to read the final value of java source directories. See `AbstractAndroidProjectHandler.preprocessVariant`.

Reply:
Oh! I meant "ad-hoc" in the sense of "NamedDomainObjectContainer" or "FileCollection" which both are live. It seems that "SourceDirectorySet" is not live this sense (like executing code the moment a new source folder was added, or letting one SourceDirectorySet extend a live other one). 

I can't see any advantage for hooking into variants instead of this "whenEvaluated", since we would in both cases just fill in the set with missing entries. 

NB: "whenEvaluated" behaves different than "afterEvaluate". It also makes sure, that all of Androids "afterEvaluate" listeners have run.

Reply:
This looks ok. Until AGP has better APIs for this, this should work perfectly fine.